### PR TITLE
feat(speechlm2): add Multi-Token Prediction (MTP) training support to SALMAutomodel

### PR DESCRIPTION
**Pull request by Slyne**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Add Multi-Token Prediction (MTP) auxiliary training support to `SALMAutomodel`, enabling the model to predict multiple future tokens per step using an auxiliary head attached to a NemotronV3 base checkpoint that ships without one.

**Collection**: speechlm2

# Changelog

- Add `_mtp_enabled` property to `SALMAutomodel` for a safe, pre-`configure_model`-safe check of whether the MTP head is attached
- Add `_build_and_attach_mtp_head()` to construct the NemotronV3 MTP head via Automodel factories and attach it as `self.llm.mtp` after the LLM is loaded (the base checkpoint has no MTP head and the normal `from_pretrained` injection path does not work for this architecture)
- FSDP2-shard the MTP head over the DP mesh in `configure_model`, since it is attached after the LLM's own FSDP wrapping
- Compute MTP auxiliary loss in the training step using calculate_mtp_loss from nemo_automodel, normalized by the same - - global token count as the main loss and scaled by `dp_size` to cancel FSDP gradient averaging
- Guard MTP against the packed-sequences (THD) path, which is not yet supported
- MTP is opt-in via `mtp.enabled: true` in the model config; relevant fields: `loss_scaling_factor`, `num_nextn_predict_layers`, `hybrid_override_pattern`, `use_repeated_layer`

# Usage
Add the below to the config file to enable MTP heads
```
# In your SALM Hydra config (e.g. conf/salm_automodel.yaml):
model:
  mtp:
    enabled: true
    num_nextn_predict_layers: 1
    loss_scaling_factor: 0.1
    hybrid_override_pattern: "*"
    use_repeated_layer: false
```
The `mtp_loss` scalar is logged on every training step (prog_bar=True) so you can monitor it in your experiment tracker alongside the main loss.

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [c](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [✗] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
This PR depends on the changes in the below PR:
- Related to https://github.com/NVIDIA-NeMo/Automodel/pull/2510

### CODE REVIEW

File: nemo/collections/speechlm2/models/salm_automodel.py [hunk 1]
```diff
@@ -900,6 +1005,14 @@ def configure_model(
...
             self._use_fsdp = True
             self.perception = fully_shard(self.perception, mesh=fsdp_mesh)
+            # The MTP head was attached AFTER the LLM's own FSDP/EP wrapping (the base
+            # checkpoint has no MTP), so it's still unsharded. Shard it over the same DP
+            # mesh — otherwise its parameters are replicated and its gradients never
+            # reduce-scatter across the DP group, so each rank would diverge on real
+            # (per-rank-different) data. The head is attention-only (no experts), so plain
+            # FSDP2 over the DP mesh is the right wrapping.
+            if self._mtp_enabled:
+                self.llm.mtp = fully_shard(self.llm.mtp, mesh=fsdp_mesh)
```
Comment (pzelasko) [inline nemo/collections/speechlm2/models/salm_automodel.py:1210]: we should guard this with an extra condition in case we finetune a model that does have MTP already in base checkpoint (e.g. Nemotron Super, future Nano models, etc)

File: nemo/collections/speechlm2/models/salm_automodel.py [hunk 2]
```diff
@@ -82,6 +82,11 @@ def device(self) -> torch.device:
                 return p._local_tensor.device if isinstance(p, DTensor) else p.device
         return super().device
 
+    @property
+    def _mtp_enabled(self) -> bool:
```
Comment (pzelasko) [inline nemo/collections/speechlm2/models/salm_automodel.py:90]: Let's make it a public property

File: nemo/collections/speechlm2/models/salm_automodel.py [hunk 3]
```diff
@@ -842,6 +922,25 @@ def configure_model(
...
+        #   * mtp_hybrid_override_pattern is read ONLY from the HF config, but the base
+        #     config class doesn't declare it, so HF drops it as an unknown kwarg;
+        #   * passing config= collides — NemotronHForCausalLM.__init__ has **kwargs, so
+        #     Automodel's _filter_kwargs_for_init can't strip the positional ``config``.
+        # So we load the LLM normally and then construct + attach the head ourselves.
+        mtp_cfg = self.cfg.get("mtp", None)
+        mtp_requested = mtp_cfg is not None and mtp_cfg.get("enabled", False)
+        if mtp_requested:
+            if self.cfg.get("packed_sequences", False):
+                raise NotImplementedError("MTP is not yet supported with packed_sequences=true (THD path).")
```
Comment (pzelasko) [inline nemo/collections/speechlm2/models/salm_automodel.py]: Can we add this before merging the PR?

File: nemo/collections/speechlm2/models/salm_automodel.py [hunk 4]
```diff
@@ -850,6 +949,12 @@ def configure_model(
             **automodel_kwargs,
         )
 
+        # Build + attach the MTP head ourselves (the base checkpoint has none).
+        if mtp_requested and not self._mtp_enabled:
+            self._build_and_attach_mtp_head(mtp_cfg, dtype)
+        if mtp_requested and not self._mtp_enabled:
+            raise RuntimeError("MTP enabled but self.llm.mtp is still None after _build_and_attach_mtp_head.")
```
Comment (pzelasko) [inline nemo/collections/speechlm2/models/salm_automodel.py:1151]: Why do we expect this to happen?

**Review by pzelasko — COMMENTED** [review]
Looks great, leaving a couple of comments

**Comment by Slyne** [comment]
Superseded by #16050, which rebases the consolidated SpeechLM2 MTP training implementation onto current main and removes the obsolete compatibility fallbacks, Automodel fork pin, and unrelated changes. Continuing review in #16050.
